### PR TITLE
Wdt 1112 appfileoverrides

### LINKED ISSUE #1112: RFE - generic file loading overrides
discoverDomain to pick up overrides and plan files when using a directory structure as described by generic file loading overrides (https://docs.oracle.com/en/middleware/fusion-middleware/weblogic-server/12.2.1.4/depgd/config.html#GUID-BBCE91E3-87ED-4D3E-A5F1-DE9F9B7F97EA).

### PR DESCRIPTION
**Pull request by CarolynRountree**
Fixes #1112 

Handle application folder directories that contain app and plan directories. 
Write directory to plan.xml config-root
make sure directory output is correct for option -remote

### CODE REVIEW

File: core/src/main/python/wlsdeploy/tool/deploy/applications_deployer.py [hunk 1]
```diff
@@ -9,10 +9,25 @@
...
 from java.io import File
 from java.io import IOException
+from java.io import FileOutputStream
+from java.lang import StringBuilder
+from java.io import BufferedReader
+from java.io import BufferedWriter
+from java.io import FileInputStream
+from java.io import FileOutputStream
+from java.io import InputStreamReader
+from java.io import OutputStreamWriter
```
Comment (rakillen) [inline core/src/main/python/wlsdeploy/tool/deploy/applications_deployer.py:19]: FileOutputStream is in the list twice.
These don't seem to be used: StringBuilder, BufferedReader, BufferedWriter, FileInputStream, InputStreamReader, OutputStreamWriter

File: core/src/main/python/wlsdeploy/tool/discover/deployments_discoverer.py [hunk 2]
```diff
@@ -2,8 +2,11 @@
 Copyright (c) 2017, 2020, Oracle Corporation and/or its affiliates.  All rights reserved.
 Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
 """
+import os
+
 from java.io import File
 from java.lang import IllegalArgumentException
+from javax.xml.parsers import DocumentBuilderFactory
```
Comment (rakillen) [inline core/src/main/python/wlsdeploy/tool/discover/deployments_discoverer.py:9]: DocumentBuilderFactory doesn't seem to be used.

File: core/src/main/python/wlsdeploy/tool/discover/deployments_discoverer.py [hunk 3]
```diff
@@ -310,14 +317,104 @@ def add_application_plan_to_archive(self, application_name, application_dict):
...
+
+    def _test_app_folder(self, source_path, plan_dir):
+        app_folder = False
+        app_dir = File(source_path).getParent()
+        if app_dir.endswith('app') and plan_dir.endswith('plan'):
+            app_folder = True
+        return app_folder
+
+    def _create_application_directory(self, application_name, application_dict):
+        _method_name = 'createAppFolder'
```
Comment (rakillen) [inline core/src/main/python/wlsdeploy/tool/discover/deployments_discoverer.py:344]: method name is _create_application_directory

**Review by rakillen — CHANGES_REQUESTED** [review]
Minor changes

**Review by rakillen — APPROVED** [review]

**Review by ddsharpe — APPROVED** [review]

File: core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java [hunk 4]
```diff
@@ -58,6 +58,12 @@ public class WLSDeployArchive {
      */
     public static final String ARCHIVE_APPS_TARGET_DIR = WLSDPLY_ARCHIVE_BINARY_DIR + "/applications";
 
+    /**
+     * Top-level archive subdirectory where the applications are stored and the subdirectory to which
+     * they will be extracted. This is for structured applications found under /app
+     */
+    public static final String ARCHIVE_APPS_FOLD_TARGET_DIR = WLSDPLY_ARCHIVE_BINARY_DIR + "/application_folder";
```
Comment (robertpatrick) [inline core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java:65]: Typically, we use plural nouns for the folders in the archive, please change to make this consistent.

File: core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java [hunk 5]
```diff
@@ -590,13 +607,41 @@ public String addApplication(String appPath) throws WLSDeployArchiveIOException
...
+        LOGGER.exiting(CLASS, METHOD, newName);
+        return newName;
+    }
+
+    public String addApplicationPlanFolder(String appName, String planDir)
+            throws WLSDeployArchiveIOException {
+        final String METHOD = "addApplicationPathFolder";
+        LOGGER.entering(CLASS, METHOD, appName, planDir);
+        File zipPlan = new File(planDir);
+        String zipprefix = ARCHIVE_APPS_FOLD_TARGET_DIR + "/" + appName + "/" + zipPlan.getName();
```
Comment (robertpatrick) [inline core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java:638]: Is there a reason to not use camel case for this variable name?

**Review by robertpatrick — CHANGES_REQUESTED** [review]

File: core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java [hunk 6]
```diff
@@ -62,7 +62,7 @@ public class WLSDeployArchive {
      * Top-level archive subdirectory where the applications are stored and the subdirectory to which
      * they will be extracted. This is for structured applications found under /app
      */
-    public static final String ARCHIVE_APPS_FOLD_TARGET_DIR = WLSDPLY_ARCHIVE_BINARY_DIR + "/application_folder";
+    public static final String ARCHIVE_APPS_FOLD_TARGET_DIR = WLSDPLY_ARCHIVE_BINARY_DIR + "/applicationsFolder";
```
Comment (robertpatrick) [inline core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java:65]: I hate to be a pain but please change this name to `applicationFolders` to more accurately reflect the purpose/content.  That is, this is a location for structured application folders not a single folder for applications.

**Review by robertpatrick — COMMENTED** [review]
Other than the one rename, the rest looks fine.  Thanks.

File: core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java (hunk 6 above)
Comment (ddsharpe) [reply core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java:65]: I think `structuredApplications` is clearer for the user.  Sounds more intuitive as to why we separated the two application types.

**Review by ddsharpe — APPROVED** [review]

**Review by robertpatrick — APPROVED** [review]

**Pull request merged**